### DISCUSSION TITLE
Support for transparency in theme colors

### DISCUSSION THREAD

**Discussion by sbdchd** [ghostty-org/ghostty#3926] [Feature Requests, Ideas]
With Terminal.app I have transparency in things like selection, so selecting looks like:

<img width="762" alt="image" src="https://github.com/user-attachments/assets/5ba00d5e-650f-4a43-936f-deab46d51aaa" />

When trying to replicate the theme I get:

<img width="592" alt="image" src="https://github.com/user-attachments/assets/66715244-4698-4a26-ab20-926fc86be45c" />

selection-background: invalid value "#ffffff30"

related: https://github.com/ghostty-org/ghostty/blob/6508fec945258f526e0adf0dcc174d954e23fb8d/src/config/Config.zig#L3751-L3753

**Comment by OmichronAgain** [comment]
Just here to add a +1 to this, especially for selection colors. I've replicated my whole theme from WezTerm to Ghostty save for the selection colors, so what follows is as accurate a comparison I can get.

---

My test setup is a Docker container with starship [default config] in a checkout of this repo (hence why the WezTerm one also says Ghostty).

WezTerm, no selection
![image](https://github.com/user-attachments/assets/1aab79da-8c61-4cf7-b4a3-65a40ecf7031)

Ghostty, no selection[^1]
![image](https://github.com/user-attachments/assets/6d4f67e8-c2c2-4ebb-a9e9-788e593de44c)

---

WezTerm, selected
![image](https://github.com/user-attachments/assets/a675c093-8772-4f82-8337-b81ce3f4b668)

Ghostty, selected (`selection-invert-fg-bg = false`)
![image](https://github.com/user-attachments/assets/0a64df6f-e36e-4eba-806e-3d2d4c1f4a4b)

Ghostty, selected (`selection-invert-fg-bg = true`)
![image](https://github.com/user-attachments/assets/9f9468ac-cd9b-4c6e-ac64-5172362f9546)

---

It doesn't seem possible to replicate the "feel" of WezTerm's selections using just "swap colors" or "use this foreground and background", as it's impossible to replicate alpha blending in a general way with only a static opaque color.

I can _mimic_ the selection color **on a black background**, but it's not the same overall:
![image](https://github.com/user-attachments/assets/b6e88846-6a8d-42cb-bd59-ac311770d8d1)
This is fragile (only works if the background hasn't been changed from black), and doesn't work for the foreground color at all.

In short: we can't replicate alpha-blended selection colors with current Ghostty options, and neither `selection-invert-fg-bg` nor the (current) propositions in #2685 can do it. To achieve this look, there would need to be support for selection colors being transparent.

In terms of the rest (i.e. not the selection colors) of the theme colors supporting transparency, it would be strange for the selection color _only_ to support it.


[^1]: Side note: evidently, WezTerm and Ghostty are using different versions of JetBrainsMono NF, but I also noticed that the `[Docker]` is dimmed in WezTerm but not in Ghostty. That particular piece is styled as `bold red dimmed` [`\033[1;2;31m`], and I'm using `bold-is-bright = true` (+the equivalent in WezTerm).
What seems to be happening is that WezTerm gives `dimmed` priority over the implied `bright` from `bold`, but Ghostty is enforcing the implied `bright`. Intuitively, it would make sense for the _explicit_ `dimmed` to override the _implicit_ `bright`, so this seems a little odd.
I've searched for "dim bold" in Issues+Discussions but haven't found anything; if this is worth more discussion, let me know and I'll open a separate item for it.

**Comment by EthanColbert8** [comment]
Also seconding this, my idea would be to have a `selection-background-opacity` that functions essentially the same way as `background-image-opacity` in v1.2.0. This would allow consistent configuration options and a lot of flexibility for the visuals for selection. Hope this is something that's feasible in Ghostty!

**Comment by mtaras0** [comment]
I would request full support for ##aarrggbb color code in a theme, for all parts of it. It would be especially helpful for high levels of background transparency, where it would be beneficial to, for example, set dimmer color to be more transparent than bright ones, rather than just darker, or for creating monochrome themes. Linear opacity interpolation for 256-color palette generation would also be absolutely fine, I think. Might produce weird effects with both bg and fg color set for some piece of output, but that's on user and theme to make sure everything works fine.
